disnake/state.py

```diff
@@ -1973,6 +1973,34 @@ def _update_guild_channel_references(self) -> None:
                 # channel will either be a TextChannel, VoiceChannel, Thread or Object
                 msg._rebind_cached_references(new_guild, channel)  # type: ignore
 
+        # these generally get deallocated once the voice reconnect times out
+        # (it never succeeds after gateway reconnects)
+        # but we rebind the channel reference just in case
+        for vc in self._voice_clients.values():
+            if not getattr(vc.channel, "guild", None):
+                continue
+
+            new_guild = self._get_guild(vc.channel.guild.id)
+            if new_guild is None:
+                continue
+
+            new_channel = new_guild._resolve_channel(vc.channel.id) or Object(id=vc.channel.id)
+            if new_channel is not vc.channel:
+                vc.channel = new_channel  # type: ignore
+
+    def _update_member_references(self) -> None:
+        messages: Sequence[Message] = self._messages or []
+        for msg in messages:
+            if not msg.guild:
+                continue
+
+            # note that unlike with channels, this doesn't fall back to `Object` in case
+            # guild chunking is disabled, but still shouldn't lead to old references being
+            # kept as `msg.author.guild` was already rebound (see above) at this point.
+            new_author = msg.guild.get_member(msg.author.id)
+            if new_author is not None and new_author is not msg.author:
+                msg.author = new_author
```

Comment:
Does this really need to be ran if the member caches are disabled? it would be impossible to get a bind to a new author, and references would still be kept to the old guild.

Reply:
The client member is always cached, regardless of cache flags or members intent, so if the member cache is disabled then this will only rebind the author of the bot's own messages.
Otherwise, the message will still point to the old bot member, as seen in this image:
![image](https://user-images.githubusercontent.com/8530778/163239379-6c799f38-e2db-4145-b6cb-79f1b2ca3600.png)

Reply:
But this sounds like a rebound message authored by the bot would have a different bound guild object than a message from the same guild, or even the message's message reference.

For example, a bot has the member cache disabled, but has the message cache enabled. They have a shard reconnect, and all messages authored by the bot are rebound to the new member. However, messages sent by other users are not rebound, as the member cache is disabled, continuing to maintain a reference to the old guild.

Reply:
> However, messages sent by other users are not rebound, as the member cache is disabled, continuing to maintain a reference to the old guild.

This is also mentioned in the comment in `_update_member_references` above.
That's where this part comes in, which runs after the guild cache is repopulated but before any chunking takes place, if enabled:
https://github.com/DisnakeDev/disnake/blob/71784c4c9f66b2a0279fe2a4495f8f7703756a72/disnake/message.py#L1120-L1121

The sequence of events is something like this:
1. reconnect shard
2. wait for guilds
3. rebind `message.guild`, `message.channel` and `message.author.guild`
4. wait for chunking to finish
5. rebind `message.author`